EIPS/eip-5388.md

```diff
@@ -0,0 +1,151 @@
+---
+eip: 5388
+title: Token-Gated HTTPS Endpoints
+description: Composable RESTful and Solidity interface to implement token-gated HTTPS endpoints usind data tokens.
+author: Tim Daubenschütz (@TimDaub)
+discussions-to: https://ethereum-magicians.org/t/token-gated-https-endpoints/10205
+status: Draft
+type: Standards Track
+category: ERC
+created: 2022-08-01
+requires: 712
+---
+
+## Abstract
+
+This standard introduces composable RESTful HTTPS and Solidity interfaces that e.g. enable paying to access HTTPS endpoints using [EIP-20](./eip-20.md) tokens.
+
+## Motivation
+
+According to RFC 7231, HTTPS status `402 Payment Required` is reserved for future use. But the future is now as Ethereum enables users to verify their identity with a server using public key infrastructure.
+
+The Ethereum community has expressed the need for a composable method of token-gating HTTPS endpoints. In this document, we outline a protocol implementable by HTTPS-providing servers to accept [EIP-20](./eip-20.md) payments prior to allow access to an authorized endpoint.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+Upon the user's initial request, a HTTPS server implementing token-gating via EIP-5388 responds with a `402 Payment Required` status code:
+
+- Its body must include a valid JSON string that complies with the Solidity Contract ABI Specification for calling contracts.
+- It must include a custom header `X-EIP-5388-CONTRACT-ADDRESS` that suggests the contract's address a call must be directed to.
```

Comment:
The use of the `X-` prefix, while very common is actually considered useless (if not harmful..) so I think it would be ok to skip it in favor `EIP-5388`.

Reply:
I think it is important to make sure this address is compatible beyond Ethereum mainnet by adding the chain id.

Reply:
> The use of the `X-` prefix, while very common is actually considered useless (if not harmful..) so I think it would be ok to skip it in favor `EIP-5388`.

Can u please add a reference that explains why the `X- prefix` is harmful/useless?

Reply:
> I think it is important to make sure this address is compatible beyond Ethereum mainnet by adding the chain id.

Makes sense, will add it.

Reply:
> Can u please add a reference that explains why the X- prefix is harmful/useless?

`X-` headers are for non-standardized headers. So if I have an Application A that needs to return a custom header (for some reason), I should prefix it with `X-`.

This is a standard. You can (and should) drop the `X-`.

Reply:
> Can u please add a reference that explains why the X- prefix is harmful/useless?

There is a good SO thread there and my good friend Peter wrote an RFC a while back https://www.rfc-editor.org/rfc/rfc6648

https://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions